cpp/src/DataStorm/NodeSessionI.cpp

```diff
@@ -175,10 +188,26 @@ NodeSessionI::destroy()
             _instance->getObjectAdapter()->remove(_publicNode->ice_getIdentity());
         }
 
-        for (const auto& [_, session] : _sessions)
+        if (auto nodeSessionManager = _nodeSessionManager.lock())
         {
             // Notify sessions of the disconnection, don't need to wait for the result.
-            session->disconnectedAsync(nullptr);
+            for (const auto& [_, session] : _sessions)
+            {
+                auto id = session->ice_getIdentity();
+                auto pos = id.name.find('-');
+                if (pos != string::npos && pos < id.name.length())
+                {
+                    // Destroy is called from the NodeSessionManager with the mutex locked.
+                    if (auto nodeSession = nodeSessionManager->getSessionNoLock(
```

Comment:
In this case could we just pass this session as a parameter on destroy? Or are there multiple layers of indirection in between?

Reply:
I don't like it very much either, but keep in mind this is all internal.